airflow/jobs/scheduler_job_runner.py

```diff
@@ -1029,6 +1048,168 @@ def _update_dag_run_state_for_paused_dags(self, session: Session = NEW_SESSION)
+        # Each scheduler should end the spans that it has started.
+        #
+        # Otel spans are designed so that only the process that starts them,
+        # has full control over their lifecycle.
+        # This also means that the process that started them, is the only one that can end them.
+        #
+        # If another scheduler has finished processing a dag_run or a task and there is a reference
+        # on the active_spans dictionary, then the current scheduler started the span,
+        # and therefore must end it.
+        dag_runs_should_end: list[DagRun] = session.scalars(
+            select(DagRun).where(DagRun.span_status == SpanStatus.SHOULD_END)
+        ).all()
+        tis_should_end: list[TaskInstance] = session.scalars(
+            select(TaskInstance).where(TaskInstance.span_status == SpanStatus.SHOULD_END)
+        ).all()
+
+        for dag_run in dag_runs_should_end:
+            active_dagrun_span = self.active_dagrun_spans.get(dag_run.run_id)
+            if active_dagrun_span is not None:
+                if dag_run.state in State.finished_dr_states:
+                    dagv = session.scalar(select(DagVersion).where(DagVersion.id == dag_run.dag_version_id))
+                    DagRun.set_dagrun_span_attrs(span=active_dagrun_span, dag_run=dag_run, dagv=dagv)
+
+                    active_dagrun_span.end(end_time=datetime_to_nano(dag_run.end_date))
+                else:
+                    active_dagrun_span.end()
+                self.active_dagrun_spans.delete(dag_run.run_id)
+                dag_run.set_span_status(status=SpanStatus.ENDED, session=session, with_commit=False)
+
+        for ti in tis_should_end:
+            active_ti_span = self.active_ti_spans.get(ti.key)
+            if active_ti_span is not None:
+                if ti.state in State.finished:
+                    self.set_ti_span_attrs(span=active_ti_span, state=ti.state, ti=ti)
+                    active_ti_span.end(end_time=datetime_to_nano(ti.end_date))
+                else:
+                    active_ti_span.end()
+                self.active_ti_spans.delete(ti.key)
+                ti.set_span_status(status=SpanStatus.ENDED, session=session, with_commit=False)
+
+    @provide_session
+    def _recreate_unhealthy_scheduler_spans_if_needed(self, dag_run: DagRun, session: Session = NEW_SESSION):
+        scheduler_health_timeout = conf.getint("scheduler", "scheduler_health_check_threshold")
+
+        # There are two scenarios:
+        #   1. scheduler is unhealthy but managed to update span_status
+        #   2. scheduler is unhealthy and didn't manage to make any updates
+        # Check the span_status first, in case the 2nd db query can be avoided (scenario 1).
+
+        # Get the latest values from the db.
+        dr: DagRun = session.scalars(
+            select(DagRun).where(
+                DagRun.run_id == dag_run.run_id,
+                DagRun.dag_id == dag_run.dag_id,
+            )
+        ).one()
```

Comment:
We "just" fetched this. If you really want to get the latest values (which I'm not sure it is worth it) then

```suggestion
        # Get the latest values from the db.
        session.refresh(dag_run)
```

Reply:
I don't think a refresh is needed. This is passed DagRun objects that come from `DagRun.get_running_dag_runs_to_examine` which gets rows via `SELECT ... FOR UPDATE` -- meaning while this transaction is active nothing else can update the rows.

(Removing this refresh also means the condition just below and at the callsite can be combined)

Reply:
> I don't think a refresh is needed. This is passed DagRun objects that come from DagRun.get_running_dag_runs_to_examine which gets rows via SELECT ... FOR UPDATE -- meaning while this transaction is active nothing else can update the rows.

You are right! I'll fix it.